File: IPWorks MQ Samples/MQTT/php/mqtt.php
<?php
/*
 * IPWorks MQ 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks MQ in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksmq
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/ipworksmq_mqtt.php');
require_once('../include/ipworksmq_const.php');
?>
<?php
global $receivedMessage;
$receivedMessage = false;
class MyMQTT extends IPWorksMQ_MQTT
{
  // Fires when a message is received
  function fireMessageIn($param) {
    global $receivedMessage;
    $receivedMessage = true;
    echo "Message from " . $param['topic'] . " at QoS " . $param['qos'] . ": \"" . $param['message'] . "\"\n";
  }
}

function input($prompt) {
  echo $prompt;
  $handle = fopen("php://stdin", "r");
  $data = trim(fgets($handle));
  fclose($handle);
  return $data;
}

$mqtt = new MyMQTT();

$host = "test.mosquitto.org";
$port = 1883;
$topic = "nsoftware/test";
$qos = 1;

try {
  echo "Connecting to " . $host . "\n";
  $mqtt->doConnectTo($host, $port);
  echo "Connected.\n";

  $input = input("Enter topic to subscribe and publish to [default: nsoftware/test]. Type 'quit' to quit.\n");
  if (strtolower($input) == "quit") return;
  if (strlen($input) > 0) $topic = $input;
  
  $mqtt->doSubscribe($topic, $qos);
  echo "Subscribed to " . $topic . ".\nEnter messages to send. Type 'quit' to quit.\n";
  
  while (true) {
    $input = input("mqtt> ");
    if (strtolower($input) == "quit") {
      echo "Quitting\n";
      return;
    } elseif (strlen($input) > 0) {
      // publish the message and wait for an incoming message
      $receivedMessage = false;
      $mqtt->doPublishMessage($topic, $qos, $input);
      while (!$receivedMessage) {
        $mqtt->doEvents();
      }
    }
  }

} catch (Exception $ex) {
  echo "Error: " . $ex->getMessage() . "\n";
}
?>

File: IPWorks MQ Samples/MQTT/java/mqtt.java
/*
 * IPWorks MQ 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks MQ in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksmq
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworksmq.*;

public class mqtt extends ConsoleDemo{

    public static void main(String[] args) {
        System.out.println("*******************************************************");
        System.out.println("* This is a demo for the IP*Works MQ MQTT Class.      *");
        System.out.println("* It prompts the user for a topic and message, then   *");
        System.out.println("* subscribes to that topic and publishes the message. *");
        System.out.println("* The demo uses a publicly available test server.     *");
        System.out.println("*******************************************************");
        MQTT mqtt = new MQTT();
        try {
            // Set up event handlers
            mqtt.addMQTTEventListener(new MQTTEventListener() {
                @Override
                public void connected(MQTTConnectedEvent mqttConnectedEvent) {

                }

                @Override
                public void connectionStatus(MQTTConnectionStatusEvent mqttConnectionStatusEvent) {

                }

                @Override
                public void disconnected(MQTTDisconnectedEvent mqttDisconnectedEvent) {

                }

                @Override
                public void error(MQTTErrorEvent mqttErrorEvent) {

                }

                @Override
                public void log(MQTTLogEvent mqttLogEvent) {

                }

                @Override
                public void messageAck(MQTTMessageAckEvent mqttMessageAckEvent) {

                }

                @Override
                public void messageIn(MQTTMessageInEvent mqttMessageInEvent) {
                    // Fires whenever a message is received
                    System.out.println(new String(mqttMessageInEvent.message)); // Print the contents of the message.
                }

                @Override
                public void messageOut(MQTTMessageOutEvent mqttMessageOutEvent) {

                }

                @Override
                public void readyToSend(MQTTReadyToSendEvent mqttReadyToSendEvent) {

                }

                @Override
                public void SSLServerAuthentication(MQTTSSLServerAuthenticationEvent mqttSSLServerAuthenticationEvent) {

                }

                @Override
                public void SSLStatus(MQTTSSLStatusEvent mqttSSLStatusEvent) {

                }

                @Override
                public void subscribed(MQTTSubscribedEvent mqttSubscribedEvent) {

                }

                @Override
                public void unsubscribed(MQTTUnsubscribedEvent mqttUnsubscribedEvent) {

                }
            });

            // Connect to the test server
            mqtt.connectTo("test.mosquitto.org", 1883);

            // Get a topic and subscribe to it
            String topic = prompt("Please enter a topic to subscribe to", ":", "nsoftware/test");

            // Subscribe at QoS Level 2
            mqtt.subscribe(topic, 2);

            Thread.sleep(1000); // Wait for a retained message

            // Get a message to post and post it to the topic
            String message = prompt("Please enter a message to post", ":", "Hello World!");
            // Publish the message at QoS level 2
            mqtt.publishMessage(topic, 2, message);

            Thread.sleep(1000); // Wait for the message to be published and returned
        } catch (Exception e) {
            displayError(e);
        }
    }
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksMQException) {
      System.out.print(" (" + ((IPWorksMQException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}




